keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+
+    ```go
+    type Issuer struct {
+        // URL points to the issuer URL in a format schema://url/path.
+        URL string
+        // CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+        CertificateAuthorityData []byte
+        // ClientID the JWT must be issued for, the "sub" field. This plugin only trusts a single
+        // client to ensure the plugin can be used with public providers.
+        // Do not affect anything with the SkipOIDCValidations option enabled.
+        // +optional
+        ClientID string
+        // SkipOIDCValidations is a flag to turn off issuer validation, client id validation.
+        // OIDC related checks.
+        // +optional
+        SkipOIDCValidations bool
+   }
+   ```
+
+2. `ClaimValidationRules` - additional rules for authorization.
+    ```go
+    type ClaimValidationRule struct {
+        // Rule is a logical expression that is written in CEL https://github.com/google/cel-go.
+        Rule string
+        // Message customize returning message for validation error of the particular rule.
+        // +optional
+        Message string
+    }
+    ```
+
+    For validation expressions, the CEL is used. They are similar to validations functions for [Custom Resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions).
+    `Rule` expression should always evaluate a boolean. Token `claims` are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
+
+    You can find a snippet of validation rules below:
+
+    ```yaml
+    claimValidationRules:
+    
+    - rule: 'claims.aud == "charmander" || claims.aud == "bulbasaur"'
+      message: clients other than charmander or bulbasaur are not allowed
+    
+    - rule: 'claims.roles.exists(r, r == "kubernetes-user")'
+      message: only kubernetes-user group members can access the cluster
+    ```
+
+3. `ClaimMappings` - rules to map claims from a token to Kubernetes user attributes.
+    ```go
+    type UserAttributes struct {
+        // Username represents an option for the username attribute.
+        Username string
+        // Groups represents an option for the groups attribute. 
+        // +optional
+        Groups string
+        // UID represents an option for the uid attribute.
+        // +optional
+        UID string
+        // Extra represents an option for the extra attribute.
+        // +optional
+        Extra []ExtraMapping
+    }
+   
+    type ExtraMapping struct {
+        // Key is a CEL expression to extract extra attribute key.
+        Key string
+        // Expression is a CEL expression to extract extra attribute value.
+        Expression string
+    }
+    ```
+
+    Every field of structures above is a CEL expression. For username and uid, the expression should evaluate a string.
```

Comment:
I'm not fluent in the problem domain enough to say where CEL expressions are most needed, but, for each of these fields, will it be more common that a string literal is supplied or that an expression is supplied?  If the expression is more common, this is the right API. If a string is more common, consider having both `Username` and `UsernameExpression` and validating that only one of the two fields is set.  If an expression will be _very_ uncommon, consider not offering CEL for that field yet. It is easier to add CEL support later than to drop it later (once GA is becomes impossible).

Reply:
I think without CEL, some cases we aimed to solve will not be covered. I like the idea that each mapping can be either a field or a CEL expression.

The API will look like this:

```go
    type UserAttributes struct {
        // Username represents an option for the username attribute.
        Username Mapping
        // Groups represents an option for the groups attribute. 
        // +optional
        Groups Mapping 
        // UID represents an option for the uid attribute.
        // +optional
        UID Mapping
        // Extra represents an option for the extra attribute.
        // +optional
        Extra []ExtraMapping
    }
   
    type ExtraMapping struct {
        // Key is a CEL expression to extract extra attribute key.
        Key Mapping
        // Value is a CEL expression to extract extra attribute value.
        Value Mapping
    }
    
    type Mapping struct {
        // Only one is possible
        Field string
        Expression string
    }
 ``` 
 
 ```yaml
 claimMappings:
   username:
      field: name
   groups:
      expression: 'claims.roles.split(",")'
 ```
 
 @enj @liggitt any thoughts on this?